### PR TITLE
Maji 3: Preact + Redux stack

### PR DESCRIPTION
**Pull request by pascalw**
This PR aggregates all the work on Maji 3.

- [x] Update project template
  - [X] Setup directory structure
  - [x] Webpack
  - [x] Babel
  - [x] Babel polyfills (+ linter conf)
  - [x] Add ESLint (base on OSS frontend)
  - [x] Sentry support
  - [x] Settings
  - ~~[ ] Initializers (bootstrap with FastClick, cordova platform-xxx support)~~
  - [x] Hello Maji demo
  - [x] Add Redux
  - [x] Add Kabisa CI support (?)
    - Dockerfile added
  - ~~[ ] Add Makefile?~~ (decided to ditch make)
  - [x] Add integration test solution (rspec, nightwatch.js?)
  - [x] Add [babel-preset-env](https://github.com/babel/babel-preset-env)
  - [x] Add https://github.com/ismay/stylelint-no-unsupported-browser-features ?
- [x] Remove all Backbone specific library code
- [x] Update Cordova Support
  - [x] Port to ES2015
  - [x] Remove jQuery dependency
- [x] Remove storage module
- [x] Basic page transition support
- [x] Update I18n support
- [x] Remove error module
- [x] Update ios_network_activity
   - [x] Port to ES2015
   - [x] Remove jQuery dependency
- [x] Fix devserver port to 9090
- [x] Add icon system (SVG)
- [x] Replace scss-lint with style-lint
- ~~[ ] Add PWA support (webpack-offline, manifest)~~ (will be addressed at later stage)
- [x] fix ios-deploy. Cordova 7 resolves NPM modules differently.
- [x] Add karma-chai-plugins (PR: https://github.com/kabisa/maji/pull/163)
- [x] Fix `bundle` not found in `bin/ci`
- [x] Fix app title no longer based on app name? (PR: https://github.com/kabisa/maji/pull/164)
- [x] Write release notes
- [x] Fix Sentry js exception due to demo "<your sentry.io public DSN here>". (PR: https://github.com/kabisa/maji/pull/162)
- [x] Fix `Unsafe attempt to load URL file://.../sprite.svg#forward-usage from frame with URL file:///.../index.html#/. 'file:' URLs are treated as unique security origins.`
- [x] Drop NPM support in CLI, only support Yarn
- [x] Exclude `cordova/` from prettier

### CODE REVIEW

**Comment by pascalw** [comment]
I've been contemplating including a Makefile again or not. I'm leaning towards not including it for the following reasons:

* `scripts` in `package.json` are a nice place to put project related scripting. Short scripts can be embedded in the package.json directly and for more complex stuff the npm script can simply refer to a bash script in `scripts/` for example.
* Makefile syntax is powerful but not very well known amongst the javascript crowd. The syntax definitely has it's warts too.
* In Maji 2 the Makefile provided a speed up for incremental builds (it could skip building if nothing changed). This is mainly useful when debugging Cordova builds so the app assets don't need to be recompiled when you're fiddling with the Cordova configuration or something like that. While that's very useful it doesn't seem to warrant having a Makefile. If anything I would like to solve this at either the build tooling level (something like `maji run --skip-build` OR solve it in Webpack).

@matthijsgroen @ruisalgado what do you think?

**Comment by ruisalgado** [comment]
@pascalw agree on avoiding make. Always felt a bit out of place and indeed Makefile syntax is not something I expect the average JS enthusiast to be familiar with. If I put myself in the shoes of someone totally new to Maji, I'd expect all the build tasks to be available through `bin/maji`

**Comment by pascalw** [comment]
Indeed `build` and `watch` are available on the maji cli too.

**Comment by matthijsgroen** [comment]
I would also go for a known setup format for JS projects, so a nice list of what is available through the `scripts` section in the package.json would be the preferred way for me.

**Comment by RoelN** [comment]
Hi, I have nothing constructive to report, except: nice going, folks!± 👍

File: project_template/webpack.config.js [hunk 1]
```diff
@@ -57,7 +57,7 @@ module.exports = {
   output: {
     path: out,
     filename: "[name].[hash].js",
-    publicPath: "./"
+    publicPath: "/"
```
Comment (pascalw) [inline project_template/webpack.config.js]: Was this change needed for nightwatch? This likely breaks in Cordova, because we need relative file based urls for Cordova.

File: project_template/README.md [hunk 2]
```diff
@@ -31,9 +31,9 @@ Prettier can also be run manually using `yarn run prettier` or `yarn run prettie
 
 ### General
 
-* Ruby + Bundler, for the integration specs
 * Node.js >=6 + NPM, for the build system
 * Chrome v59.x or higher to run the tests headless. You might need install it from the [beta channel](https://www.google.com/chrome/browser/beta.html).
+* Java 8 (or higher) to run the end-to-end tests (it powers the selenium server)
```
Comment (pascalw) [inline project_template/README.md]: When I was playing with Nightwatch I used Chromedriver directly. That way we don't have a dependency on Java/Selenium.

File: project_template/webpack.config.js (hunk 1 above)
Comment (ruisalgado) [inline project_template/webpack.config.js]: the relative path broke the `historyApiFallback` on the dev server, which I guess will also cause issues with Nightwatch if u want to jump straight to a particular URL when starting a test

File: project_template/README.md (hunk 2 above)
Comment (ruisalgado) [inline project_template/README.md]: yeah I noticed that was an option but then we have to consciously make the decision that we don't care about running the tests against other browsers. I didn't care so much about the Java dependency because presumably people will use this to make their cross platform apps, at which point they need JDK for android

File: project_template/webpack.config.js (hunk 1 above)
Comment (pascalw) [inline project_template/webpack.config.js]: Hm ok but then we still need to make sure that this doesn't break Cordova, can you double check that? (`yarn run build && open dist/index.html` should be enough)

File: project_template/README.md (hunk 2 above)
Comment (pascalw) [inline project_template/README.md]: Right, fair point.

File: project_template/webpack.config.js (hunk 1 above)
Comment (ruisalgado) [inline project_template/webpack.config.js]: yup I'll look into it

**Comment by ruisalgado** [comment]
@pascalw made some changes to the directory structure by putting all tests under the same /test dir. Also rolled out our own dev server instead of using the one provided by webpack - this server is used for both `yarn start` and `yarn test:e2e`. It could be pulled upstream into the maji lib, although it's quite convenient to have it in the project template if u wanna customize it (for example adding `historyApiFallback`)

**Comment by ruisalgado** [comment]
@pascalw could you publish this to NPM tagged as `beta`? It'll make it a LOT more convenient to bootstrap new maji 3 projects

**Comment by ruisalgado** [comment]
fyi beta3 has been published to npm

**Comment by pascalw** [comment]
@RoelN can you look into this issue with SVG icons?

* Build with `yarn build`
* `open dist/index.html`
* SVG icons are not loaded and Chrome console displays:
`Unsafe attempt to load URL file://.../sprite.svg#forward-usage from frame with URL file:///.../index.html#/. 'file:' URLs are treated as unique security origins.`

File: project_template/package.json [hunk 3]
```diff
@@ -1,52 +1,81 @@
...
   "scripts": {
-    "test": "bin/ci",
-    "test:unit": "APP_ENV=test karma start --single-run",
-    "test:integration": "APP_ENV=test PRE_BUILT=true bundle exec rspec",
-    "test:watch": "APP_ENV=test karma start --watch",
-    "start": "make watch",
-    "build": "make dist"
+    "build": "NODE_ENV=${NODE_ENV:-production} webpack",
+    "start": "LIVERELOAD=${LIVERELOAD:-true} script/dev-server.js",
+    "test": "NODE_ENV=${NODE_ENV:-test} karma start config/karma.js --single-run",
```
Comment (pascalw) [inline project_template/package.json]: I noticed that `yarn test` is only running unit tests now. Would it make sense to have `yarn test` delegate to `bin/ci` so all tests are ran (unit, integration, linting etc). We could add `test:unit` task then to run karma tests.

/cc @ruisalgado @matthijsgroen

Comment (ruisalgado) [reply project_template/package.json]: agreed... given the conventions around `yarn/npm test`, makes sense to have it run all the available tests

**Comment by pascalw** [comment]
@matthijsgroen @ruisalgado I've fixed one of the remaning issues in the todo list. Only thing left now is to remove npm support, but I'm fine leaving that for a later release. (Though that technically would mean Maji 4 as it would be breaking).

**Comment by matthijsgroen** [comment]
don't we need to move #161 in as well?

**Comment by pascalw** [comment]
@matthijsgroen yes, merged! 👍

**Comment by pascalw** [comment]
Jenkins test this please

**Comment by pascalw** [comment]
🎉
